core/src/outputs.rs

```diff
@@ -0,0 +1,130 @@
+
+/// Output container for Miden VM programs.
+///
+/// Miden program outputs contain the full state of the stack at the end of execution as well as the
+/// addresses in the overflow table which are required to reconstruct the table (when combined with
+/// the overflow values from the stack state).
+///
+/// `stack` is expected to be ordered as if they elements were popped off the stack one by one.
+/// Thus, the value at the top of the stack is expected to be in the first position, and the order
+/// of the rest of the output elements will also match the order on the stack.
+///
+/// `overflow_addrs` is expected to start with the `prev` address value from the first row in the
+/// overflow table (the row representing the deepest element in the stack) and then be followed by
+/// the address (`clk` value) of each row in the table starting from the deepest element in the
+/// stack and finishing with the row which was added to the table last.
+
+#[derive(Debug, Clone, Default)]
+pub struct ProgramOutputs {
+    /// The elements on the stack at the end of execution.
+    stack: Vec<u64>,
+    /// The overflow table row addresse required to reconstruct the final state of the table.
+    overflow_addrs: Vec<u64>,
+}
+
+impl ProgramOutputs {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    pub fn new(stack: Vec<u64>, overflow_addrs: Vec<u64>) -> Self {
+        Self {
+            stack,
+            overflow_addrs,
+        }
+    }
+
+    pub fn from_elements(stack: Vec<Felt>, overflow_addrs: Vec<Felt>) -> Self {
+        let stack = stack.iter().map(|&v| v.as_int()).collect::<Vec<_>>();
+        let overflow_addrs = overflow_addrs
+            .iter()
+            .map(|&v| v.as_int())
+            .collect::<Vec<_>>();
+
+        Self {
+            stack,
+            overflow_addrs,
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the stack outputs, which is state of the stack at the end of execution converted to
+    /// integers.
+    pub fn stack(&self) -> &[u64] {
+        &self.stack
+    }
+
+    /// Returns the number of requested stack outputs or returns the full stack if fewer than the
+    /// requested number of stack values exist.
+    pub fn stack_outputs(&self, num_outputs: usize) -> &[u64] {
+        if num_outputs < self.stack.len() {
+            return &self.stack[..num_outputs];
+        }
+
+        &self.stack
+    }
+
+    /// Returns an iterator of the elements that were on the top of the stack at the end of
+    /// execution.
+    pub fn stack_top_iter(&self) -> Box<dyn Iterator<Item = Felt> + '_> {
+        Box::new(
+            self.stack
+                .iter()
+                .take(MIN_STACK_DEPTH)
+                .map(|v| Felt::new(*v)),
+        )
+    }
```

Comment:
There are a few places where we need an iterator over the stack top (as field elements). I was trying to avoid iterating over the whole thing twice every time (once to convert to Felt and once to use it). Is there a better way to return an iterator so that we just iterate once (converting to field elements and then continuing with the other iteration steps)?

Reply:
Not sure there is a better way. But we call this method only in a handful of places, and we iterate over only 16 elements. So, I wouldn't worry about performance implications. I'd rather have a bit simpler code.